### ISSUE TITLE
Mongolian words are not rendered correctly (not connected)

### ISSUE THREAD

**Issue by mongol-surah** [sile-typesetter/sile#1726]
A Mongolian word, if preceded by a Latin character, is not rendered correctly. The characters are not connected.
<img width="104" alt="image" src="https://user-images.githubusercontent.com/118827982/223881595-bc23b819-cf0e-4bc0-b524-48d0e149d35b.png">

Sile source code:
```
\begin[papersize=a5,direction=TTB-LTR]{document}
\bidi-off
\nofolios
\language[main=mon]
\font[filename=mnglwhiteotf.ttf,size=12pt]
A ᠵᠢᠷᠤᠮᠲᠤ  
\end{document}
````

Could you give some hints about the problem? Thank you!

Version: SILE v0.14.8 (LUA 5.4)
Platform: WSL Ubuntu; SILE installed via homebrew

PS: the font I used is [this](http://www.mongolfont.com/cn/font/mnglwhiteotf.html)

**Comment by mongol-surah** [comment]
It turns out that, for a piece of multi-lingual text, one has to break it into parts according to languages, then send the parts separately into harfbuzz. See the discussion [here](https://github.com/harfbuzz/harfbuzz/discussions/4156).

**Comment by khaledhosny** [comment]
SILE should be doing the itemization, so if this is not working it is a SILE bug.

**Comment by Mulliken** [comment]
> SILE should be doing the itemization, so if this is not working it is a SILE bug.

Hi Khalid! Thank you for pointing this out! Do you happen to know which part of SILE may cause the problem? I'd like to dig in a little, but reading the whole source code and navigate to the part related to the issue is a bit overwhelming for me. Thank you!!!

**Comment by mongol-surah** [comment]
> SILE should be doing the itemization, so if this is not working it is a SILE bug.

I ran some examples with the `-d tokenizaer`  flag. The debug information indicates that SILE doesn't tokenize Mongolian texts very well currently. See the discussion https://github.com/sile-typesetter/sile/discussions/1728.

**Comment by mongol-surah** [comment]
The source code related to itemization (or tokenization as SILE calls it) is the following two pieces. I will continue reading the code and hopefully, I could identify where the problem lies

https://github.com/sile-typesetter/sile/blob/c8a1467494b6cbcf3582c56598cb68f55c11df83/languages/unicode.lua#L15
https://github.com/sile-typesetter/sile/blob/c8a1467494b6cbcf3582c56598cb68f55c11df83/shapers/base.lua#L106

**Comment by khaledhosny** [comment]
> Do you happen to know which part of SILE may cause the problem?

No, unfortunately.

**Comment by Omikhleia** [comment]
@mongol-surah Look how `shaper:createNnodes` is implemented (in `shapers/base.lua`):

Debugging here, it is passed the _unshaped nodes_ (containing the input, i.e. your mixed latin and mongolian characters at this point). It is _first_ shaped (with Harfbuzz, so hitting the issue you and Khaled mentioned), _prior_ to any kind of (ICU-based) tokenization (splitting spaces, word-breaks, etc. ... not sure it's expected to do more than that...).

So it seems to me that SILE doesn't perform any itemization automatically _before_ shaping.

You could enforce it manually. On way (in SIL syntax) is to use a group `{ }` around e.g. the latin bits.

```
\begin[papersize=a5,direction=TTB-LTR]{document}
\bidi-off
\nofolios
\language[main=mn]
\font[filename=mnglwhiteotf.ttf,size=12pt]
{A} ᠵᠢᠷᠤᠮᠲᠤ
\par
\end{document}
```

It tedious, I know...

Another way would be to use an inputfilter to to the splitting for you, based on character ranges.

Both of these might be workaround for a deeper problem/bug, I don't know. Maybe this is just how SILE is expected to behave ^^

**Comment by Omikhleia** [comment]
Quick PoC for the inputfilter workaround:
```
\begin[papersize=a5,direction=TTB-LTR]{document}
\bidi-off
\nofolios
\lua{
-- Mongolian is U+1800..U+18AF
local m1 = SU.utf8charfromcodepoint("U+1800")
local m2 = SU.utf8charfromcodepoint("U+18AF")

local mnSplitFilter = function (node, content, options)
 if type(node) == "table" then return node end
 local result = {}
 for token in SU.gtoke(node, "["..m1.."-"..m2.."]+") do
 if token.string then
    result[#result+1] = token.string
   else
    result[#result+1] = token.separator
   end
 end
 return result
end

local class = SILE.documentState.documentClass
class:loadPackage("inputfilter")
class:registerCommand("mongolian", function (_, content)
  local result = class.packages.inputfilter:transformContent(content, mnSplitFilter)
  SILE.process(result)
end)
}

\language[main=mn]
\font[filename=mnglwhiteotf.ttf,size=12pt]
\begin{mongolian}%
A ᠵᠢᠷᠤᠮᠲᠤ
\end{mongolian}
\par
\end{document}
```

**Comment by khaledhosny** [comment]
Probably related to https://github.com/sile-typesetter/sile/issues/83, I think the Arabic text works accidentally because bidi will put the Arabic text in a different token, but any mixed scripts with same direction do not work as there is no automatic script itemization.

**Comment by mongol-surah** [comment]
> 

Thanks, Omikhleia! Thank you for the workarounds and for telling me where the problem lies! I will continue to investigate this issue. I think SILE could and should do something like what Pango does - itemization according to the language, font, direction and other properties.

**Comment by Omikhleia** [comment]
@khaledhosny is right, the general limitation is the one addressed in #83 --> Closing this issue as Duplicate.
